cleanrl/muesli_atari_envpool_async_jax_scan_impalanet_machado.py

```diff
@@ -0,0 +1,1395 @@
+            episodic_life=False,  # Hessel et al. 2022, Muesli paper, Table 4
+            repeat_action_probability=0.25,  # Hessel et al. 2022, Muesli paper, Table 4
+            noop_max=1,  # Hessel et al. 2022, Muesli paper, Table 4
+            full_action_space=True,  # Hessel et al. 2022, Muesli paper, Table 4
+            max_episode_steps=int(
+                108000 / 4
+            ),  # Hessel et al. 2022, Muesli paper, Table 4, we divide by 4 because of the skipped frames
+            reward_clip=True,
+            seed=seed,
+        )
+        envs.num_envs = num_envs
+        envs.single_action_space = envs.action_space
+        envs.single_observation_space = envs.observation_space
+        envs.is_vector_env = True
+        return envs
+
+    return thunk
+
+
+class ResidualBlock(nn.Module):
+    channels: int
+
+    @nn.compact
+    def __call__(self, x):
+        inputs = x
+        x = nn.relu(x)
+        x = nn.Conv(
+            self.channels,
+            kernel_size=(3, 3),
+        )(x)
+        x = nn.relu(x)
+        x = nn.Conv(
+            self.channels,
+            kernel_size=(3, 3),
+        )(x)
+        return x + inputs
+
+
+class ConvSequence(nn.Module):
+    channels: int
+
+    @nn.compact
+    def __call__(self, x):
+        x = nn.Conv(
+            self.channels,
+            kernel_size=(3, 3),
+        )(x)
+        x = nn.max_pool(x, window_shape=(3, 3), strides=(2, 2), padding="SAME")
+        x = ResidualBlock(self.channels)(x)
+        x = ResidualBlock(self.channels)(x)
+        return x
+
+
+class RepresentationNetwork(nn.Module):
+    """Large IMPALA network.
+
+    See Espeholt et al. 2018, IMPALA paper, Figure 4.
+    """
+
+    action_dim: tp.Sequence[int]
+    channelss: tp.Sequence[int] = (16, 32, 32)
+    lstm_dim: int = 256
+    reward_clip: int = 1
+
+    @nn.compact
+    def __call__(self, carry, x, last_reward, last_action):
+        """Encode a batch of observations into the latent space.
+
+        Args:
+            carry: The carry to pass to the LSTM. <(batch_size, lstm_dim), (batch_size, lstm_dim)>
+            x: The batch of observations. (batch_size, channels, height, width)
+            last_reward: The reward from the previous step. (batch_size)
+            last_action: The action from the previous step. (batch_size)
+        """
+        x = jnp.transpose(x, (0, 2, 3, 1))
+        x = x / 255.0
+        for channels in self.channelss:
+            x = ConvSequence(channels)(x)
+        x = nn.relu(x)
+        batch_size = x.shape[0]
+        x = x.reshape(batch_size, -1)
+        x = nn.Dense(256, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0))(x)
+        x = nn.relu(x)
+        # Reward clipping (See Espeholt et al. 2018, IMPALA paper, Table C.3. Used for single levels.
+        # DMLab-30 uses a different reward transformation: see Figure C.1.)
+        last_reward = jnp.clip(last_reward, -self.reward_clip, self.reward_clip)
```

Comment:
This won't be necessary because `envs = envpool.make(reward_clip=True)` already does this.

Reply:
Thanks! I'll update that!

Reply:
Actually, Muesli does advantage clipping elsewhere. Also, non-linear transformations already handle reward scaling in the predicted reward and value models ( $\hat{r_k}$ and $\hat{v_k}$). I'll turn off reward clipping in the environment and leave the reward clipping to the representation network.

Reply:
That makes sense!